ext/java/org/msgpack/jruby/Factory.java

```diff
@@ -0,0 +1,133 @@
+package org.msgpack.jruby;
+
+
+import org.jruby.Ruby;
+import org.jruby.RubyClass;
+import org.jruby.RubyObject;
+import org.jruby.RubyArray;
+import org.jruby.RubyHash;
+import org.jruby.RubyIO;
+import org.jruby.RubyInteger;
+import org.jruby.RubyFixnum;
+import org.jruby.RubyString;
+import org.jruby.RubySymbol;
+import org.jruby.runtime.builtin.IRubyObject;
+import org.jruby.anno.JRubyClass;
+import org.jruby.anno.JRubyMethod;
+import org.jruby.runtime.ThreadContext;
+import org.jruby.runtime.ObjectAllocator;
+import org.jruby.util.ByteList;
+
+import static org.jruby.runtime.Visibility.PRIVATE;
+
+@JRubyClass(name="MessagePack::Factory")
+public class Factory extends RubyObject {
+  private Ruby runtime;
+  private Packer.ExtRegistry packerExtRegistry;
+  private Unpacker.ExtRegistry unpackerExtRegistry;
+
+  public Factory(Ruby runtime, RubyClass type) {
+    super(runtime, type);
+    this.runtime = runtime;
+  }
+
+  static class FactoryAllocator implements ObjectAllocator {
+    public IRubyObject allocate(Ruby runtime, RubyClass type) {
+      return new Factory(runtime, type);
+    }
+  }
+
+  @JRubyMethod(name = "initialize")
+  public IRubyObject initialize(ThreadContext ctx) {
+    this.packerExtRegistry = new Packer.ExtRegistry(ctx.getRuntime());
+    this.unpackerExtRegistry = new Unpacker.ExtRegistry(ctx.getRuntime());
+    return this;
+  }
+
+  public Packer.ExtRegistry packerRegistry() {
+    return this.packerExtRegistry.dup();
+  }
+
+  public Unpacker.ExtRegistry unpackerRegistry() {
+    return this.unpackerExtRegistry.dup();
+  }
+
+  @JRubyMethod(name = "packer", optional = 1)
+  public Packer packer(ThreadContext ctx, IRubyObject[] args) {
+    return Packer.newPacker(ctx, packerRegistry(), args);
+  }
+
+  @JRubyMethod(name = "unpacker", optional = 1)
+  public Unpacker unpacker(ThreadContext ctx, IRubyObject[] args) {
+    return Unpacker.newUnpacker(ctx, unpackerRegistry(), args);
+  }
+
+  @JRubyMethod(name = "registered_types_internal", visibility = PRIVATE)
+  public IRubyObject registeredTypesInternal(ThreadContext ctx) {
+    // unpacker
+    Unpacker.ExtRegistry reg = unpackerRegistry();
+    RubyHash mapping = RubyHash.newHash(ctx.getRuntime());
+    for (int i = 0; i < 256; i++) {
+      if (reg.array[i] != null) {
+        mapping.fastASet(RubyFixnum.newFixnum(ctx.getRuntime(), i - 128), reg.array[i]);
+      }
+    }
+
+    IRubyObject[] ary = { packerRegistry().hash, mapping };
+    return RubyArray.newArray(ctx.getRuntime(), ary);
+  }
+
+  @JRubyMethod(name = "register_type", required = 2, optional = 1)
+  public IRubyObject registerType(ThreadContext ctx, IRubyObject[] args) {
+    // register_type(type, Class)
+    // register_type(type, Class, packer: proc-like, unpacker: proc-like)
```

Comment:
I assume these comments are some kind of type declaration to show which calls can be made, but that should be API documentation, or even better shown with tests.